flatwrap.sh

```diff
@@ -0,0 +1,109 @@
+  echo
+  echo "A temporary directory named SYSEXTNAME in the current folder will be created and deleted again."
+  echo "All files in the sysext image will be owned by root."
+  echo "To use a different architecture than amd64 pass 'ARCH=arm64' as environment variable (current value is '${ARCH}')."
+  "${SCRIPTFOLDER}"/bake.sh --help
+  exit 1
+fi
+
+FOLDER="$1"
+SYSEXTNAME="$2"
+shift
+shift
+PATHS=("$@")
+
+if [ "${ETCMAP}" = host ]; then
+  HOST+=" /etc"
+else
+  CHROOT+=" /etc"
+fi
+if [ "${VARMAP}" = host ]; then
+  HOST+=" /var"
+else
+  CHROOT+=" /var"
+fi
+if [ "${HOMEMAP}" = host ]; then
+  HOST+=" /home"
+else
+  CHROOT+=" /home"
+fi
+
+rm -rf "${SYSEXTNAME}"
+mkdir -p "${SYSEXTNAME}/usr/local/${SYSEXTNAME}"
+mkdir -p "${SYSEXTNAME}/usr/local/${SYSEXTNAME}/mount-dir"
+
+cp -ar "${FOLDER}/." "${SYSEXTNAME}/usr/local/${SYSEXTNAME}"
+for DIR in ${CHROOT} ${HOST}; do
+  # Required for the unpriv unshare case
+  mkdir -p "${SYSEXTNAME}/usr/local/${SYSEXTNAME}/${DIR}"
+done
+
+CARGS=()
+BWCARGS=()
+for DIR in ${CHROOT}; do
+  CARGS+=(mkdir -p "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/mount-dir/${DIR}" "&&" mount --bind "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/${DIR}" "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/mount-dir/${DIR}" "&&")
+  BWCARGS+=(--bind "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/${DIR}" "${DIR}")
+done
+HARGS=()
+BWHARGS=()
+for DIR in ${HOST}; do
+  HARGS+=(mkdir -p "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/mount-dir/${DIR}" "&&" mount --bind "${DIR}" "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/mount-dir/${DIR}" "&&")
+  BWHARGS+=(--bind "${DIR}" "${DIR}")
+done
+
+tee "${SYSEXTNAME}/usr/local/${SYSEXTNAME}/unshare-helper" > /dev/null <<EOF
+#!/bin/sh
+mount -t tmpfs tmpfs "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/mount-dir" && ${CARGS[@]} ${HARGS[@]} chroot "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/mount-dir" "\$@"
+EOF
+chmod +x "${SYSEXTNAME}/usr/local/${SYSEXTNAME}/unshare-helper"
+
+for ENTRY in "${PATHS[@]}"; do
+  DIR=$(dirname "${ENTRY}")
+  mkdir -p "${SYSEXTNAME}/${DIR}"
+  if [ -L "${FOLDER}/${ENTRY}" ]; then
+    TARGET=$(realpath -m --relative-base="${FOLDER}" "${FOLDER}/${ENTRY}")
+    ln -fs "/usr/local/${SYSEXTNAME}/${TARGET}" "${SYSEXTNAME}/${ENTRY}"
+  elif [ -d "${FOLDER}/${ENTRY}" ] || [ ! -x "${FOLDER}/${ENTRY}" ]; then
+    NAME=$(basename "${ENTRY}")
+    ln -fs --no-target-directory "/usr/local/${SYSEXTNAME}/${ENTRY}" "${SYSEXTNAME}/${DIR}/${NAME}"
+  else
+      tee "${SYSEXTNAME}/${ENTRY}" > /dev/null <<EOF
+#!/bin/sh
+if [ "\$(id -u)" = 0 ]; then
+  exec unshare -m "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}/unshare-helper" "${ENTRY}" "\$@"
+elif command -v bwrap >/dev/null && [ "\${NOBWRAP}" = "" ]; then
+  exec bwrap ${BWCARGS[@]} ${BWHARGS[@]} "${ENTRY}" "\$@"
+else
+  exec unshare -m -U -c --root "\${FLATWRAP_ROOT-}/usr/local/${SYSEXTNAME}" "${ENTRY}" "\$@"
```

Comment:
I noticed that unpriv ovelay mounts would work to emulate the bind mount. Read-only case would be `mount -t overlayfs -t overlay overlay -o lowerdir=/source:/somethingempty /target`. The read-only case would also be a fallback when the write case fails because that needs to have a writable working dir on the same filesystem as the source `mount -t overlayfs -t overlay overlay -o lowerdir=/somethingempty,upperdir=/source,workdir=/path-on-source-fs /target` (The work dir must not contain the upperdir and vice versa, and for the work dir a `work` sub directory will get created).

Since in the unpriv. case the command wouldn't have write permissions anyway for `/var` or `/etc` it should be ok to use the read-only fallback mount. For `/home` it's a bit special because it normally contains the users's home which is writable but the rest is not, and one could special case mounting `$HOME` but even then one needs a writable work dir that is not in `$HOME`. `/var/tmp` might be ok if it's not a tmpfs. What could work is mapping the subfolders of `$HOME` at the cost of having to create a work dir for each like `$HOME/.cache/SUBFOLDER-work`. I think I would go with a read-only mount for the unpriv. case, maybe add a fresh tmpfs for `/tmp` and just require `bwrap` for anything more special.

Reply:
Alternative idea: Don't do the chroot step but prepare the original `/` as needed, i.e., bind mount the read-only sysext contents over `/usr` and other things from `CHROOT` while we either ignore HOST or kick out unwanted dirs by mounting an empty dir over it (needs two empty different dirs at hand in the sysext).

The advantage is that the host dirs stay in place as wanted but a challenge is that the mount command runs "live" and the final unshare command, too, so that one might need to keep the a reference to the host binaries around: mount host `/usr` to `/host/usr` and then run `/host/usr/lib/ld-linux-x86-64.so.2 --library-path /host/… mount -t …` and same for `unshare --map-user=…`.

Reply:
Turns out I was doing it wrong and should use `mount --rbind` instead of `mount --bind`. Then the user namespace can do bind mounts. Will update the code.

Reply:
Done: When bwrap is not available things should work the same through the unshare commands. I don't even know if there is a benefit of using bwrap because this also seems to lose the mapping of other users.